### PR TITLE
refactor fetch_all_coins

### PR DESCRIPTION
**Pull request by NicolaLS**
Refactored the `mint::fetch_all_coins` a little bit, expose also `CoinFinalizationData` in `list_active_issuances()` and refactor `ln-gateway::background_fetch()` a bit

### CODE REVIEW

**Comment by NicolaLS** [comment]
Also changed the `list_active_issuances()` a bit. I need that for something and also for the future it dosn't hurt to expose more information, callers can just drop the other thing if they don't need it

**Comment by elsirion** [comment]
Had a call about this PR today, waiting for discussed changes before further review.

**Comment by NicolaLS** [comment]
just noticed my IDE seems to restructure the imports..sry will see what I can do about that

**Comment by NicolaLS** [comment]
fixed the formating stuff @elsirion

File: client/client-lib/src/mint/mod.rs [hunk 1]
```diff
@@ -1,4 +1,4 @@
-mod db;
+pub mod db;
```
Comment (NicolaLS) [inline client/client-lib/src/mint/mod.rs:1]: ..maybe this should not be pub..then I'd have to remove the intra doc-link in `gateway.rs` though

File: ln-gateway/src/lib.rs [hunk 2]
```diff
@@ -83,7 +83,7 @@ impl LnGateway {
         Ok(preimage)
     }
 
-    #[instrument(skip_all, fields(%contract_id))]
+    #[instrument(skip_all, fields(% contract_id))]
```
[the remark this answers is not in the document]
Comment (NicolaLS) [reply ln-gateway/src/lib.rs:86]: yes. will remove

File: ln-gateway/src/lib.rs [hunk 3]
```diff
@@ -157,21 +157,14 @@ async fn background_fetch(federation_client: Arc<GatewayClient>, _ln_client: Arc
...
-                async move {
-                    if let Err(e) = federation_client.fetch_coins(out_point).await {
-                        debug!(error = %e, "Fetching coins failed");
-                    }
-                }
-            })
-            .collect::<Vec<_>>();
-        futures::future::join_all(pending_fetches).await;
-        minimint_api::task::sleep_until(least_wait_until).await;
+        let active_issuances = federation_client.list_fetchable_coins();
```
[the remark this answers is not in the document]
Comment (NicolaLS) [reply ln-gateway/src/lib.rs:160]: yeah now that you point that out, `fetch_all_coins` now returns an empty Vec immediately  if there are no active issuances. will refactor that again

File: ln-gateway/src/lib.rs [hunk 4]
```diff
@@ -157,21 +157,14 @@ async fn background_fetch(federation_client: Arc<GatewayClient>, _ln_client: Arc
...
+        if active_issuances.is_empty() {
+            minimint_api::task::sleep_until(least_wait_until).await;
+            continue;
+        }
+        if let Err(e) = federation_client.fetch_all_coins().await {
```
[the remark this answers is not in the document]
Comment (NicolaLS) [reply ln-gateway/src/lib.rs:165]: Agreed, but we avoid cloning the fed_client and also its just way slimmer. I asked @justinmoon about that he said that I should go for `fetch_all_coins`but I can change that if you wan't (I think it dosnt do much performance wise though)

File: client/client-lib/src/mint/mod.rs [hunk 5]
```diff
@@ -216,36 +216,38 @@ impl<'c> MintClient<'c> {
...
+        let active_issuances = self.list_active_issuances();
+        if active_issuances.is_empty() {
+            return Ok(Vec::new());
+        }
 
-        // TODO: return out points instead
-        let mut tx_ids = vec![];
-        for (OutputFinalizationKey(out_point), _) in active_issuances {
+        let mut out_points = vec![];
+        for (active_out_point, _) in active_issuances {
```
[the remark this answers is not in the document]
Comment (NicolaLS) [reply client/client-lib/src/mint/mod.rs:243]: mhm yes that's a very good idea

[the remark this answers is not in the document]
Comment (Maan2003) [reply client/client-lib/src/mint/mod.rs:243]: actually `collect` doesn't poll futures concurrently.

we will have to use `futures::join_all` or `FuturesUnordered`.

**Comment by NicolaLS** [comment]
thank you @Maan2003 for the extensive help on the concurrent spending.

File: ln-gateway/src/lib.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (NicolaLS) [reply ln-gateway/src/lib.rs]: solved by concurrent fetching

File: ln-gateway/src/lib.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (NicolaLS) [reply ln-gateway/src/lib.rs:160]: solved by `Ok(o) if o.is_empy => ..` arm

File: ln-gateway/src/lib.rs [hunk 6]
```diff
@@ -157,21 +157,14 @@ async fn background_fetch(federation_client: Arc<GatewayClient>, _ln_client: Arc
...
-                    }
-                }
-            })
-            .collect::<Vec<_>>();
-        futures::future::join_all(pending_fetches).await;
-        minimint_api::task::sleep_until(least_wait_until).await;
+        match federation_client.fetch_all_coins().await {
+            Ok(o) if o.is_empty() => {
+                minimint_api::task::sleep_until(least_wait_until).await;
+                continue;
```
Comment (Maan2003) [inline ln-gateway/src/lib.rs:163]: this continue is useless, we are at end of iteration anyways

File: ln-gateway/src/lib.rs [hunk 7]
```diff
@@ -157,21 +157,14 @@ async fn background_fetch(federation_client: Arc<GatewayClient>, _ln_client: Arc
...
+            }
+            Err(e) => debug!(error = %e, "Fetching coins failed"),
+            _ => continue,
```
Comment (Maan2003) [inline ln-gateway/src/lib.rs:166]: same here

**Review by Maan2003 — APPROVED** [review]
LGTM, just minor nits

**Comment by NicolaLS** [comment]
just ran `bash scripts/final-tests.sh` with no problem but I saw the CI failed for some reason ?..will be not available until monday unfortunately

**Comment by elsirion** [comment]
This bug is weird, it only happens on the merged version of your branch, not the branch itself. You should probably rebase on master and see why these two tests get stuck.

File: ln-gateway/src/lib.rs [hunk 8]
```diff
@@ -157,21 +157,13 @@ async fn background_fetch(federation_client: Arc<GatewayClient>, _ln_client: Arc
...
-                }
-            })
-            .collect::<Vec<_>>();
-        futures::future::join_all(pending_fetches).await;
-        minimint_api::task::sleep_until(least_wait_until).await;
+        match federation_client.fetch_all_coins().await {
+            Ok(o) if o.is_empty() => {
+                minimint_api::task::sleep_until(least_wait_until).await;
+            }
+            Err(e) => debug!(error = %e, "Fetching coins failed"),
```
Comment (Maan2003) [inline ln-gateway/src/lib.rs:164]: we should also sleep in error case, otherwise it spams the server with a lot of request.

**Comment by NicolaLS** [comment]
I messed up the rebase, i'll just rewrite it..

**Pull request closed without merging**